### PR TITLE
APM Traces for Jenkins Build/Pipelines

### PR DESCRIPTION
**Pull request by drodriguezhdez**
###  DISCLAIMER
This PR adds an experimental feature in alpha state.
Internal use only.

### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

- Added the logic to send APM traces of Jenkins Builds/Pipelines using Datadog Agent.

### Description of the Change

The `DatadogTraceBuildLogic` class implements the logic of sending the build traces. During the execution of the `onStarted` method, a new span that represents the Jenkins Build is created. This span is finished in the `onComplete` method. The `BuildSpanManager` class is used to propagate the same span between A method and the B method. This is required because it's not possible to store the complete active Span in a Jenkins Action. Additionally, the `SpanContext` of the Jenkins Build span is injected in the `BuildSpanAction`. This logic is invoked from the `DatadogBuildListener` class.

The `DatadogTracePipelineLogic` class implements the logic of sending the pipeline traces. This logic is only executed in the las `FlowNode` of the pipeline. The idea is, based on that last node, to recreate the full pipeline tree with the complete information of each node, and then, to send the spans which represent that tree. The `BuildPipeline` and `BuildPipelineNode` classes have been implemented to build the Jenkins Pipeline nodes tree. To correlate the Jenkins Build with the pipeline, the `SpanContext` is extracted from the `BuildSpanAction`. This logic is invoked from the `DatadogGraphListener` class.

The `DatadogStepListener` class implements the logic of extract the important information of each `StepContext`, and store them in the `StepDataManager` using the `StepDescriptor` as key. This information will be available during the construction of the node spans in the `DatadogTracePipelineLogic` execution. 

To create and send the traces is used the [APM Java Tracer API](https://github.com/DataDog/dd-trace-java). 
This logic is only available for the `Datadog Agent` mode in the Jenkins Plugin. 

Finally, the `DatadogGlobalConfiguration` class and the UI template has been modified to support the configuration of the `Trace Collection Port` input and `Enable Trace Collection` flag.

### Verification Process

Apart from the addition of the unit and integration tests to this PR, the Jenkins plugin with this feature is available in the [Jenkins instance of the CI-App team](http://172.29.196.157:8080/). 

Check the [APM Traces view in Datadog](https://app.datadoghq.com/apm/trace/818761293892439615?spanID=7521683232818409527&env=ciapp-sandbox&sort=time&colorBy=service&graphType=flamegraph&shouldShowLegend=true) corresponding to this [Jenkins pipeline execution](http://172.29.196.157:8080/job/dotnet-hello-world/job/master/60/).

### Additional Notes

You can find further information of this feature in the [CI/CD observability technical design](https://docs.google.com/document/d/1F50jzFbTEWckfvrti7oHtOFYNMHY6gIWAaaCkifrlfg/edit?ts=5f0f4a84#heading=h.9d29ab5ksdcr) document.

Finally, do not hesitate to join to [#ci-app](https://dd.slack.com/archives/C016LJLRZQ9) Slack channel if you have more questions.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    private String nodeName;
+    private String nodeHostname;
+    private AnnotatedLargeText logText;
+    private long startTime;
+    private long startTimeMicros;
+    private long endTime;
+    private long endTimeMicros;
+    private String result;
+    private boolean error;
+    private Throwable errorObj;
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java:72]: Same comment as above but with `error` and `errorObj`

File: src/test/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracesUtilities.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    public static final String SAMPLE_STEP_END_NODE_NAME = "stepEndNodeName";
+
+    public static final String SAMPLE_STEP_ATOM_NODE_ONE_ID = "stepAtomNodeOneId";
+    public static final String SAMPLE_STEP_ATOM_NODE_ONE_NAME = "stepAtomNodeOneName";
+    public static final String SAMPLE_STEP_ATOM_NODE_TWO_ID = "stepAtomNodeTwoId";
+    public static final String SAMPLE_STEP_ATOM_NODE_TWO_NAME = "stepAtomNodeTwoName";
+
+    private static Map<String, FlowNode> flowNodeById;
+
+    public static Map<String, FlowNode> getDummyPipeline() {
```
Comment (sarah-witt) [inline src/test/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracesUtilities.java:47]: Could you also write a test for a trace for a job/`FreestyleProject`?

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogGlobalConfiguration.java [hunk 3]
```diff
@@ -332,6 +342,20 @@ public FormValidation doCheckTargetLogCollectionPort(@QueryParameter("targetLogC
         return FormValidation.ok("Valid Log Collection Port");
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogGlobalConfiguration.java:345]: I think the option only is shown when `Use the Datadog Agent to report to Datadog (recommended)` is selected.

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 4]
```diff
@@ -604,6 +622,182 @@ public static String getJenkinsUrl() {
...
+        }
+        return flowNode.getAction(LabelAction.class) != null;
+    }
+
+    /**
+     * Returns true if a {@code FlowNode} is a Pipeline node.
+     * @param flowNode
+     * @return flag indicating if a flowNode is a Pipeline node.
+     */
+    public static boolean isPipelineNode(FlowNode flowNode) {
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:721]: As far as I understand the [Pipeline syntax](https://www.jenkins.io/doc/book/pipeline/syntax/) doc, there are nodes which represent Pipelines, Stages, and Steps. Following the same approach as the method `isStageNode(...)`, this method returns if the evaluated node is the node that represents the "pipeline" node itself.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DogStatsDClient.java [hunk 5]
```diff
@@ -213,6 +229,32 @@ private boolean reinitializeLogger(boolean force) {
...
+            return true;
+        }
+
+        if(!DatadogUtilities.getDatadogGlobalDescriptor().isCollectBuildTraces() || this.traceCollectionPort == null) {
+            return false;
+        }
+
+        logger.info("Re/Initialize Datadog-Plugin Tracer: hostname = " + this.hostname + ", traceCollectionPort = " + this.traceCollectionPort);
+        final DDTracer.DDTracerBuilder tracerBuilder = DDTracer.builder();
+        //NOTE Enable traceCollectionPort when APM Java Tracer 0.59.0 was released.
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DogStatsDClient.java:249]: Yes. Updated!

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java [hunk 6]
```diff
@@ -646,4 +647,10 @@ private String getJenkinsVersion(){
         }
         return this.jenkinsVersion;
     }
+
+    @Override
+    public Tracer tracer() {
+        throw new UnsupportedOperationException("");
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java:653]: Absolutely. Changed!

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    private NodeType type;
+    private boolean internal;
+    private Map<String, Object> args = new HashMap<>();
+    private Map<String, String> envVars = new HashMap<>();
+    private String workspace;
+    private String nodeName;
+    private String nodeHostname;
+    private AnnotatedLargeText logText;
+    private long startTime;
+    private long startTimeMicros;
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java:67]: In the context of OpenTracing, the Spans have its `startTime` and `endTime` in microseconds, but the API methods don't indicate this fact. Using the wrong time unit is a common issue when you are working with OTracing, especially in those languages that do not support microseconds (like Java).

For that reason, I consider having an extra `getter` indicating the microsecond unit clearer to remember that the default unit of `startTime` and `endTime` is not microseconds.

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java (hunk 1 above)
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java:72]: In this context, the `error` flag can be true but `errorObj` can be null when the status `ERROR` is propagated to the parent nodes.

Added some comments to bring more light to this fact.

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanAction.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ */
+public class BuildSpanAction extends InvisibleAction {
+
+    private Map<String, String> buildSpanPropatation;
+
+    protected BuildSpanAction(){
+        this.buildSpanPropatation = new HashMap<>();
+    }
+
+    public static BuildSpanAction newAction() {
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanAction.java:19]: Sure. Changed 👍

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import org.datadog.jenkins.plugins.datadog.clients.ClientFactory;
+import org.datadog.jenkins.plugins.datadog.model.BuildData;
+import org.datadog.jenkins.plugins.datadog.model.BuildPipelineNode;
+
+import java.util.logging.Logger;
+
+/**
+ * Keeps the logic to send traces related to Jenkins Build.
+ */
+public class DatadogTraceBuildLogic {
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java:18]: Changed 👍 

Created the following methods in `DatadogClient`:
```
    void startBuildTrace(BuildData buildData, Run run);
    void finishBuildTrace(BuildData buildData);
    void sendPipelineTrace(Run<?, ?> run, FlowNode flowNode);
```

Finally, the accessor for `Tracer` has been removed from `DatadogClient`.

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Keeps the logic to send traces related to Jenkins Pipelines.
+ */
+public class DatadogTracePipelineLogic {
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java:26]: Changed! 👍 

Now, this logic is used inside `DatadogClient`.

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+        final BuildSpanAction buildSpanAction = run.getAction(BuildSpanAction.class);
+        if(buildSpanAction == null) {
+            return;
+        }
+
+        final FlowEndNode flowEndNode = (FlowEndNode) flowNode;
+        final BuildPipeline pipeline = new BuildPipeline();
+        final DepthFirstScanner scanner = new DepthFirstScanner();
+        List<FlowNode> currentHeads = flowEndNode.getExecution().getCurrentHeads();
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java]: Done! Added some comments.
Take a look at it and check if it is clearer now 👍

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DogStatsDClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DogStatsDClient.java]: Yes, in the previous version of the Java tracer (0.58.0) there was a bug when the tracer was built.

Now, version 0.59.0 is used and the bug has been fixed.

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipeline.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents a Jenkins Pipeline.
+ * To represent the stages is used an n-ary tree.
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipeline.java]: Changed! Thanks!

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
+package org.datadog.jenkins.plugins.datadog.traces;
+
+import io.opentracing.Span;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Used to propagate the build Span between onStart() and onComplete() methods.
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java:9]: I'd say it's not possible to create our own implementation of Span, cause the Datadog Span maintains the correct logic to be used in the Datadog Tracer.

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 14]
```diff
@@ -604,6 +622,182 @@ public static String getJenkinsUrl() {
         }
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java]: Moved those methods into the classes where they are being used.

File: src/test/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracesUtilities.java (hunk 2 above)
Comment (drodriguezhdez) [reply src/test/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracesUtilities.java:46]: Added! 👍

**Comment by drodriguezhdez** [comment]
Updated `DEVELOPMENT` and `README` documentation.

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogGlobalConfiguration.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogGlobalConfiguration.java:351]: Seems good enough, as long as enabling it doesn't break the Http client the warning message seems totally fine 👍

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java:70]: I agree with Sarah, for that use case it's probably best not to store both startTime and startTimeMicros as two different floats. Providing multiple getters is fine, but I'm afraid having the same information in two different place will lead in the future into updating one but forgetting to update the other inside that class.

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java:9]: I'm not sure I understand your comment. Is there anything preventing to create a class DatadogSpan such as
```java
import io.opentracing.Span;

public class DatadogSpan extends Span implements Serializable {
    private void writeObject(java.io.ObjectOutputStream stream) throws IOException {
        // write the span to stream
    }

    private void readObject(java.io.ObjectInputStream stream) throws IOException, ClassNotFoundException {
        // read the span from the stream
    }

}
```

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java:19]: Great, love it 👍

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java:9]: Yeah, I'd say it's not possible to do that.

In this case `Span` is an interface that is implemented by all internal classes of the APM Java Tracer which contains the whole logic to send the traces.

[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java:9]: Oh I didn't realize `Span` was an interface, fair enough 👍

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java:70]: But if we have multiple getters, one for milliseconds and other for microseconds, where do you suggest to have the logic of multiplying by 1000 to obtain microseconds, in the getter method? Are we comfortable with add this kind of logic in a getter?

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 15]
```diff
@@ -604,6 +622,58 @@ public static String getJenkinsUrl() {
...
+        if (flowNode == null) {
+            return false;
+        }
+        if (flowNode.getAction(StageAction.class) != null) {
+            // Legacy style stage block without a body
+            // https://groups.google.com/g/jenkinsci-users/c/MIVk-44cUcA
+            return true;
+        }
+        if (flowNode.getAction(ThreadNameAction.class) != null) {
+            // TODO comment
```
[the remark this answers is not in the document]
Comment (drodriguezhdez) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:662]: I think any change in this method should come from `master` branch, as it was just moved to `DatadogUtilities`. 
What do you think? The original method is [here](https://github.com/jenkinsci/datadog-plugin/blob/a4a3eb615dd70bc8af6b5ae59973b1ab810506c1/src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogGraphListener.java#L161).

[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:662]: my bad, didn't realize that

**Review by FlorianVeaux — APPROVED** [review]
Great 💯

**Comment by drodriguezhdez** [comment]
@sarah-witt did you have the opportunity of taking a look at my comments?

**Review by FlorianVeaux — APPROVED** [review]

**Pull request merged**